src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Compilation.cs

```diff
@@ -308,11 +308,20 @@ public ISymbolNode ComputeConstantLookup(ReadyToRunHelperId lookupKind, object t
                     {
                         var type = (TypeDesc)targetOfLookup;
 
-                        // We counter-intuitively ask for a constructed type symbol. This is needed due to IDynamicInterfaceCastable.
-                        // If this cast happens with an object that implements IDynamicIntefaceCastable, user code will
-                        // see a RuntimeTypeHandle representing this interface.
                         if (type.IsInterface)
-                            return NodeFactory.MaximallyConstructableType(type);
+                        {
+                            // We counter-intuitively ask for a constructed type symbol. This is needed due to IDynamicInterfaceCastable.
+                            // If this cast happens with an object that implements IDynamicInterfaceCastable, user code will
+                            // see a RuntimeTypeHandle representing this interface.
+                            if (NodeFactory.DevirtualizationManager.CanHaveDynamicInterfaceImplementations(type))
+                            {
+                                return NodeFactory.MaximallyConstructableType(type);
+                            }
+                            else
+                            {
+                                return NecessaryTypeSymbolIfPossible(type);
```

Comment:
Would this ever return a Necessary type symbol? The scanning phase saw the MaximallyConstructible type for this so I think this would no longer be allowed to downgrade. I think this file can be left as it was before this PR and have no impact on anything.

Reply:
I believe it will. Based on the diff sizes in the rt-sz run, most of the savings I had in the initial PR I think come from here (as I didn't have the interface dispatch cell work done yet).

Reply:
In particular, the casting operation would have introduced a dependency on the MaximallyConstructible type for this, but it wouldn't have marked it. I assume that the downgrade isn't blocked if it isn't marked.

Reply:
Yeah I took a look at the reruns of rt-sz and you're right. Only case I saw that actually optimized it away on the mtsat files was the interface dispatch cell.